src/cpu.rs

```diff
@@ -12,15 +12,237 @@
+            allow(dead_code)
+        )]
+        mask: u32,
+
+        #[cfg_attr(not(target_os = "ios"), allow(dead_code))]
+        ios: bool,
+    }
+
+    impl Feature {
+        #[inline(always)]
+        pub fn available(&self, _: super::Features) -> bool {
+            #[cfg(all(target_os = "ios", any(target_arch = "arm", target_arch = "aarch64")))]
+            {
+                return self.ios;
+            }
+
+            #[cfg(all(
+                any(target_os = "android", target_os = "linux"),
+                any(target_arch = "arm", target_arch = "aarch64")
+            ))]
+            {
+                return self.mask == self.mask & unsafe { GFp_armcap_P };
+            }
+
+            #[cfg(not(any(target_arch = "arm", target_arch = "aarch64")))]
+            {
+                return false;
+            }
+        }
+    }
+
+    // Keep in sync with `ARMV7_NEON`.
+    #[cfg(any(target_arch = "aarch64", target_arch = "arm"))]
+    pub(crate) const NEON: Feature = Feature {
+        mask: 1 << 0,
+        ios: true,
+    };
+
+    // Keep in sync with `ARMV8_AES`.
+    pub(crate) const AES: Feature = Feature {
+        mask: 1 << 2,
+        ios: true,
+    };
+
+    // Keep in sync with `ARMV8_PMULL`.
+    pub(crate) const PMULL: Feature = Feature {
+        mask: 1 << 5,
+        ios: true,
+    };
+
+    #[cfg(all(
+        any(target_os = "android", target_os = "linux"),
+        any(target_arch = "arm", target_arch = "aarch64")
+    ))]
+    extern "C" {
+        static mut GFp_armcap_P: u32;
+    }
+}
+
+#[cfg_attr(
+    not(any(target_arch = "x86", target_arch = "x86_64")),
+    allow(dead_code)
+)]
+pub(crate) mod intel {
+    pub(crate) struct Feature {
+        word: usize,
+        mask: u32,
+    }
+
+    impl Feature {
+        #[inline(always)]
+        pub fn available(&self, _: super::Features) -> bool {
+            #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+            {
+                extern "C" {
+                    static mut GFp_ia32cap_P: [u32; 4];
+                }
+                return self.mask == self.mask & unsafe { GFp_ia32cap_P[self.word] };
+            }
+
+            #[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+            {
+                return false;
+            }
+        }
+    }
+
+    pub(crate) const FXSR: Feature = Feature {
+        word: 0,
+        mask: 1 << 24,
+    };
+
+    pub(crate) const PCLMULQDQ: Feature = Feature {
+        word: 1,
+        mask: 1 << 1,
+    };
+
+    pub(crate) const SSSE3: Feature = Feature {
+        word: 1,
+        mask: 1 << 9,
+    };
+
+    #[cfg(target_arch = "x86_64")]
+    pub(crate) const MOVBE: Feature = Feature {
+        word: 1,
+        mask: 1 << 22,
+    };
+
+    pub(crate) const AES: Feature = Feature {
+        word: 1,
+        mask: 1 << 25,
+    };
+
+    #[cfg(target_arch = "x86_64")]
+    pub(crate) const AVX: Feature = Feature {
+        word: 1,
+        mask: 1 << 28,
+    };
```

Comment:
For SGX, assuming we're going to use `RDRAND`, we'll need to augment this with a new 

```rust
    pub(crate) const RDRAND: Feature = Feature {
        word: 1,
        mask: 1 << 30,
    };
```